include/dem/ray_tracing.h

```diff
@@ -0,0 +1,230 @@
+
+  /**
+   * @brief Print information about the photons that have been inserted during an
+   * insertion time step.
+   *
+   * @param[in] inserted_photon Number of inserted photon at the start of the
+   * simulation.
+   * @param[in] pcout Printing in parallel
+   */
+  void
+  print_insertion_info(const unsigned int       &inserted_photon,
+                       const ConditionalOStream &pcout);
+
+  /**
+   * @brief Insert particles and photons at the beginning of a ray tracing
+   * simulation.
+   */
+  void
+  insert_particles_and_photons();
+
+  /**
+   * @brief Generate the output file of the particle ray tracing in parallel.
+   *
+   * @param[in] points Vector of points containing every intersection points
+   * between photons and particles.
+   */
+  void
+  write_output_results(const std::vector<Point<dim>> &points);
+
+  /**
+   * @brief Execute the last post-processing at the end of the simulation and
+   * output test results if necessary.
+   *
+   * @param[in] intersection_points Vector of points containing every
+   * intersection between photons and particles.
+   */
+  void
+  finish_simulation(std::vector<Point<3>> &intersection_points);
+
+  /**
+   * @brief Execute the ray tracing algorithm to find intersection points between
+   * photons and particles.
+   *
+   * @tparam move_photon Boolean to indicate if the photon should move at the
+   * end of the loop. This is set to true when the loop is done on the local
+   * cell neighboring list.
+   *
+   * @param[in] cell_list Data structure containing the local or ghost
+   * neighboring cell to each local cell in the triangulation.
+   * @param[in,out] photon_intersection_points_map A map containing information
+   * about each intersection point found during the ray tracing. The key of the
+   * map is the particle index of the photon. The value is a tuple containing
+   * the distance between the initial location of the photon and the
+   * intersection point, the intersection point and an iterator to the photon
+   * that needs to be removed.
+   */
+  template <bool move_photon>
+  void
+  find_intersection(
+    typename dem_data_structures<dim>::cells_neighbor_list &cell_list,
+    ankerl::unordered_dense::map<
+      types::particle_index,
+      std::tuple<double, Point<dim>, Particles::ParticleIterator<dim>>>
+      &photon_intersection_points_map);
+
+  /**
+   * @brief The MPI communicator used for the parallel simulation.
+   */
+  MPI_Comm mpi_communicator;
+
+  /**
+   * @brief The number of MPI processes used for the parallel simulation.
+   */
+  const unsigned int n_mpi_processes;
+
+  /**
+   * @brief The rank of the current MPI process.
+   */
+  const unsigned int this_mpi_process;
+
+  /**
+   * @brief The output stream used for the parallel simulation, only used by
+   * the process 0.
+   */
+  ConditionalOStream pcout;
+
+  /**
+   * @brief The parameters of the Ray Tracing simulation.
+   */
+  RayTracingSolverParameters<dim> parameters;
```

Comment:
Since this is passed at the constructor,  I think this should be const. Generally simulation parameters should not change during the simulation.